##### To implement the Polynomial Linear Regression using Employee Salary Data


##### Step 1:  Import all the necessary libraries

In [185]:
import  numpy               as  np
import  pandas              as  pd
import  matplotlib.pyplot   as  plt
import  seaborn             as  sns

from    sklearn.model_selection   import  train_test_split, GridSearchCV, StratifiedKFold
from    sklearn.preprocessing     import  StandardScaler, PolynomialFeatures
from    sklearn.linear_model      import  LinearRegression
from    sklearn.metrics           import  mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score

##### Step 2:  Load the Dataset

In [186]:
df = pd.read_csv("Employee_Salary.csv")

In [187]:
df.head()

,YearsExperience,Salary
0,1.1,39343.0
1,1.3,46205.0
2,1.5,37731.0
3,2.0,43525.0
4,2.2,39891.0


#### OBSERVATIONS:

1.  The above dataset explains about the salaries earned by the employees based on the years of experience.

##### Step 3: Perform the Data Exploration

In [188]:
#### get the top five rows of the dataset

df.head()

,YearsExperience,Salary
0,1.1,39343.0
1,1.3,46205.0
2,1.5,37731.0
3,2.0,43525.0
4,2.2,39891.0


In [189]:
#### get the last five rows of the dataset

df.tail()

,YearsExperience,Salary
25,9.0,105582.0
26,9.5,116969.0
27,9.6,112635.0
28,10.3,122391.0
29,10.5,121872.0


In [190]:
#### get the total length of the dataset

print("Total length of the dataset is:", len(df))

Total length of the dataset is: 30


In [191]:
#### Shape of the dataset 

print("Shape of the dataset is:", df.shape)

Shape of the dataset is: (30, 2)


In [192]:
##### get the columns used in the dataset

df.columns

Index(['YearsExperience', 'Salary'], dtype='object')

In [193]:
#### get the information about the columns used in the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   YearsExperience  30 non-null     float64
 1   Salary           30 non-null     float64
dtypes: float64(2)
memory usage: 608.0 bytes


#### OBSERVATIONS:

1.  Both the YearsExperience and Salary are floating in nature.

In [194]:
#### get the descriptive summary statistics about the columns in the dataset

df.describe()

,YearsExperience,Salary
count,30.000000,30.000000
mean,5.313333,76003.000000
std,2.837888,27414.429785
min,1.100000,37731.000000
25%,3.200000,56720.750000
50%,4.700000,65237.000000
75%,7.700000,100544.750000
max,10.500000,122391.000000


#### OBSERVATIONS:

1.  It explains about the descriptive summary statistics about each numerical columns used in the dataset.

#### Step 4: Perform Data Cleaning

In [195]:
#### To check for the NULL records in the dataset

df.isnull().sum()

YearsExperience    0
Salary             0
dtype: int64

#### OBSERVATIONS:

1.  There are no NULL records in the dataset.

In [196]:
#### To check for the duplicate records used in the dataset

df[df.duplicated()]

,YearsExperience,Salary


#### OBSERVATIONS:

1.  There are no duplicate records in the dataset.

#### Step 5: Divide the dataset into independent and dependent variables

In [197]:
#### independent variables

X = df.drop(columns='Salary',axis=1)

print(X)

    YearsExperience
0               1.1
1               1.3
2               1.5
3               2.0
4               2.2
5               2.9
6               3.0
7               3.2
8               3.2
9               3.7
10              3.9
11              4.0
12              4.0
13              4.1
14              4.5
15              4.9
16              5.1
17              5.3
18              5.9
19              6.0
20              6.8
21              7.1
22              7.9
23              8.2
24              8.7
25              9.0
26              9.5
27              9.6
28             10.3
29             10.5


In [198]:
##### dependent variables

Y = df[['Salary']]

print(Y)

      Salary
0    39343.0
1    46205.0
2    37731.0
3    43525.0
4    39891.0
5    56642.0
6    60150.0
7    54445.0
8    64445.0
9    57189.0
10   63218.0
11   55794.0
12   56957.0
13   57081.0
14   61111.0
15   67938.0
16   66029.0
17   83088.0
18   81363.0
19   93940.0
20   91738.0
21   98273.0
22  101302.0
23  113812.0
24  109431.0
25  105582.0
26  116969.0
27  112635.0
28  122391.0
29  121872.0


#### Step 6: Divide the independent and dependent variables into training and testing data

In [199]:
from  sklearn.model_selection  import  train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=42)

In [200]:
X_train

,YearsExperience
28,10.3
24,8.7
12,4.0
0,1.1
4,2.2
16,5.1
5,2.9
13,4.1
11,4.0
22,7.9


In [201]:
X_test

,YearsExperience
27,9.6
15,4.9
23,8.2
17,5.3
8,3.2
9,3.7


In [202]:
print("Shape of the input training data is:", X_train.shape)
print("Shape of the input testing  data is:", X_test.shape)

Shape of the input training data is: (24, 1)
Shape of the input testing  data is: (6, 1)


In [203]:
Y_train

,Salary
28,122391.0
24,109431.0
12,56957.0
0,39343.0
4,39891.0
16,66029.0
5,56642.0
13,57081.0
11,55794.0
22,101302.0


In [204]:
Y_test

,Salary
27,112635.0
15,67938.0
23,113812.0
17,83088.0
8,64445.0
9,57189.0


In [205]:
print("Shape of the output training data is:", Y_train.shape)
print("Shape of the output testing  data is:", Y_test.shape)

Shape of the output training data is: (24, 1)
Shape of the output testing  data is: (6, 1)


#### Step 7: Perform Feature Scaling on the inputs

In [206]:
from  sklearn.preprocessing  import  StandardScaler

##### create an object for Standard Scaler

sc = StandardScaler()

##### using the object for Standard Scaler, transform the inputs

X_train_scaled = sc.fit_transform(X_train)

X_test_scaled = sc.transform(X_test)

In [207]:
X_train_scaled

array([[ 1.77445774],
       [ 1.21912622],
       [-0.41216011],
       [-1.41869849],
       [-1.03690807],
       [-0.03036969],
       [-0.79395053],
       [-0.37745189],
       [-0.41216011],
       [ 0.94146047],
       [-1.34928205],
       [-1.27986561],
       [ 1.32325088],
       [-1.10632451],
       [ 0.66379471],
       [ 1.49679198],
       [ 0.24729607],
       [ 1.84387418],
       [ 0.55967005],
       [-0.68982587],
       [-0.44686833],
       [-0.23861901],
       [ 0.28200429],
       [-0.75924231]])

In [208]:
X_test_scaled

array([[ 1.5315002 ],
       [-0.09978613],
       [ 1.04558513],
       [ 0.03904675],
       [-0.68982587],
       [-0.51628477]])

#### OBSERVATIONS:

1.  Here all the input features have been scaled in the same scale range.

==================== (a.)  LINEAR REGRESSION ======================================

#### Step 8: Train the Linear Regression Model

In [209]:
from  sklearn.linear_model  import  LinearRegression

#### create an object for linear regression

lin = LinearRegression()

#### using the object for linear regression, train the model

lin.fit(X_train_scaled, Y_train)

LinearRegression()

#### Step 9: Predict the Model

In [210]:
#### Predict the model

Y_pred_lin = lin.predict(X_test_scaled)

In [211]:
Y_pred_lin

array([[115790.21011287],
       [ 71498.27809463],
       [102596.86866063],
       [ 75267.80422384],
       [ 55477.79204548],
       [ 60189.69970699]])

#### Step 10: Find the slope and intercept of the model

In [212]:
print("Slope of the model is:", lin.coef_[0])
print("Intercept of the model is:", lin.intercept_)

Slope of the model is: [27151.53742583]
Intercept of the model is: [74207.625]


#### Step 11: Find the accuracy of the model

In [213]:
#### Find the accuracy of the model

ac = r2_score(Y_test, Y_pred_lin)

print("R2 Score of the model is:", ac)

R2 Score of the model is: 0.9024461774180498


==================== (a.)  POLYNOMIAL REGRESSION ======================================

#### Step 12: Perform the polynomial degrees

In [214]:
from  sklearn.preprocessing import PolynomialFeatures


##### create the object for Polynomial degrees

poly = PolynomialFeatures(degree=4)


##### using the object for Polynomial degrees, transform the inputs

X_train_poly = poly.fit_transform(X_train)

print(X_train_poly)

[[1.00000000e+00 1.03000000e+01 1.06090000e+02 1.09272700e+03
  1.12550881e+04]
 [1.00000000e+00 8.70000000e+00 7.56900000e+01 6.58503000e+02
  5.72897610e+03]
 [1.00000000e+00 4.00000000e+00 1.60000000e+01 6.40000000e+01
  2.56000000e+02]
 [1.00000000e+00 1.10000000e+00 1.21000000e+00 1.33100000e+00
  1.46410000e+00]
 [1.00000000e+00 2.20000000e+00 4.84000000e+00 1.06480000e+01
  2.34256000e+01]
 [1.00000000e+00 5.10000000e+00 2.60100000e+01 1.32651000e+02
  6.76520100e+02]
 [1.00000000e+00 2.90000000e+00 8.41000000e+00 2.43890000e+01
  7.07281000e+01]
 [1.00000000e+00 4.10000000e+00 1.68100000e+01 6.89210000e+01
  2.82576100e+02]
 [1.00000000e+00 4.00000000e+00 1.60000000e+01 6.40000000e+01
  2.56000000e+02]
 [1.00000000e+00 7.90000000e+00 6.24100000e+01 4.93039000e+02
  3.89500810e+03]
 [1.00000000e+00 1.30000000e+00 1.69000000e+00 2.19700000e+00
  2.85610000e+00]
 [1.00000000e+00 1.50000000e+00 2.25000000e+00 3.37500000e+00
  5.06250000e+00]
 [1.00000000e+00 9.00000000e+00 8.100000

In [215]:
X_test_poly = poly.transform(X_test)

print(X_test_poly)

[[1.0000000e+00 9.6000000e+00 9.2160000e+01 8.8473600e+02 8.4934656e+03]
 [1.0000000e+00 4.9000000e+00 2.4010000e+01 1.1764900e+02 5.7648010e+02]
 [1.0000000e+00 8.2000000e+00 6.7240000e+01 5.5136800e+02 4.5212176e+03]
 [1.0000000e+00 5.3000000e+00 2.8090000e+01 1.4887700e+02 7.8904810e+02]
 [1.0000000e+00 3.2000000e+00 1.0240000e+01 3.2768000e+01 1.0485760e+02]
 [1.0000000e+00 3.7000000e+00 1.3690000e+01 5.0653000e+01 1.8741610e+02]]


#### OBSERVATIONS:

1.  Here all the input polynomial features have been scaled in the same scale range.

#### Step 13: Train the Polynomial Linear Regression

In [216]:
from  sklearn.linear_model  import  LinearRegression

#### create an object for linear regression

lin_poly = LinearRegression()

#### using the object of linear regression, train the model

lin_poly.fit(X_train_poly, Y_train)

LinearRegression()

#### Step 14: Predict the Model

In [217]:
#### Predict the model

Y_pred_poly = lin_poly.predict(X_test_poly)

In [218]:
Y_pred_poly

array([[116127.90403692],
       [ 70521.87794064],
       [105073.1106405 ],
       [ 75036.10503207],
       [ 52734.46641057],
       [ 57552.8996866 ]])

#### Step 15: Find the slope and intercept of the model

In [219]:
print("Slope of the model is:"    , lin_poly.coef_[0])
print("Intercept of the model is:", lin_poly.intercept_)

Slope of the model is: [    0.         -6990.88091369  4106.28106878  -379.61541823
    11.34499537]
Intercept of the model is: [44306.59622763]


#### Step 16: Find the accuracy of the model

In [220]:
#### Find the accuracy of the model

ac_poly = r2_score(Y_test, Y_pred_poly)

print("R2 Score of the model is:", ac_poly)

R2 Score of the model is: 0.9029791879053799


#### OBSERVATIONS:

1.  The R2 Score of the model using linear regression is 0.9024

2.  The R2 Score of the model using Polynomial linear regression is 0.9029

3.  The R2 Score (Polynomial linear) is almost same as the linear regression regression.

##### Step 17: Perform Hyper Parameter Tuning using Grid Search CV

In [221]:
#### Create the pipeline

from  sklearn.linear_model  import  Ridge, Lasso, ElasticNet

from  sklearn.pipeline  import  Pipeline

pipeline = Pipeline([
    ("sc"            ,   StandardScaler())                        ,
    ("ridge"         ,   Ridge(alpha = 0.01))
])

In [222]:
#### Create the parameter  grid

param_grid       =        {
    "ridge__alpha"        :       [0.01,    0.1,    1,  10, 100]
}

In [223]:
#### perform the Startified K-Fold Cross Validation

from  sklearn.model_selection  import  KFold


cv = KFold(n_splits = 5, shuffle = True, random_state = 42)

In [224]:
#### perform the Grid Search CV

grid = GridSearchCV(
    pipeline                                     ,
    param_grid   =    param_grid                 ,
    cv           =        cv                     ,
    scoring      =     'r2'                     
)

In [225]:
#### Train the grid search cv

grid.fit(X_train_poly, Y_train)

GridSearchCV(cv=KFold(n_splits=5, random_state=42, shuffle=True),
             estimator=Pipeline(steps=[('sc', StandardScaler()),
                                       ('ridge', Ridge(alpha=0.01))]),
             param_grid={'ridge__alpha': [0.01, 0.1, 1, 10, 100]},
             scoring='r2')

In [226]:
#### perform the best parameters and the estimators of the model

best_parameters = grid.best_params_

print("Best parameters of the model is:", best_parameters)



best_estimator = grid.best_estimator_

print("Best estimator of the model is:", best_estimator)

Best parameters of the model is: {'ridge__alpha': 0.01}
Best estimator of the model is: Pipeline(steps=[('sc', StandardScaler()), ('ridge', Ridge(alpha=0.01))])


In [227]:
#### predict the model

Y_pred_grid_poly = best_estimator.predict(X_test_poly)

In [228]:
Y_pred_grid_poly

array([[116876.85367413],
       [ 70003.06233539],
       [105315.14876475],
       [ 74258.27776125],
       [ 53673.70676766],
       [ 58112.68417563]])

In [229]:
#### Find the accuracy of the model


ac_poly = r2_score(Y_test, Y_pred_grid_poly)

print("R2 Score of the model is:", ac_poly)


R2 Score of the model is: 0.9056073712811729


#### OBSERVATIONS:

1.  The R2 Score of the model using linear regression is 0.9024

2.  The R2 Score of the model using Polynomial linear regression is 0.9029

3.  The R2 Score (Polynomial linear) is almost same as the linear regression regression.

4.  The R2 Score of the Polynomial Linear regression using Grid Search CV is 0.9056 which is slightly higher than linear and Polynomial Linear regression.